### ISSUE TITLE
Standalone build does not include server-side variables from `.env.local`

### ISSUE THREAD

**Issue by evakkuri** [vercel/next.js#46296]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

  Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:04:44 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T8103
    Binaries:
      Node: 18.13.0
      npm: 8.19.3
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.1.7-canary.28
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

Standalone mode (output: "standalone")

### Link to the code that reproduces this issue

https://github.com/evakkuri/nextjs-standalone-envvar-bug

### To Reproduce

- Build Docker image: `docker build -t nextjs-standalone-envvar-bug:latest .`
  - Dockerfile from NextJS [with-docker](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile) example, apart from creating `.env` and `.env.local` files 
- Run container: `docker run -p 3000:3000 nextjs-standalone-envvar-bug `
- Navigate to `http://localhost:3000`
- Note that from server logs that `process.env.API_URL_1` is undefined, and all other environment variables are defined as expected

### Describe the Bug

Standalone build does not include server-side environment variables from `.env.local`, at least with the Dockerfile as per NextJS example.

### Expected Behavior

Also `process.env.API_URL_1`should have retrieved a value from `.env.local`, indicating that server-side variables are read at build-time from `.env.local`.

### Which browser are you using? (if relevant)

Firefox 110.0 (64-bit)

### How are you deploying your application? (if relevant)

Dockerfile (included in repo)

**Comment by feledori** [comment]
Having this issue at the moment. Have you found a workaround?

**Comment by evakkuri** [comment]
I have not unfortunately. In my case I luckily could get by with using just .env

**Comment by feledori** [comment]
Yes I did the same on before build i copy files to .env and temp delete the other files. This is also a affects .env.test.

**Comment by klawson6** [comment]
I have experienced this issue - public variables from .env.local are inlined into the build perfectly fine, but the server-side variables are missing from the resultant .env file which is bundled with the build. 

The workaround I've found is to simply copy the .env.local file to the build folder. In the case of your repo, copying to the app folder of the runner stage in Dockerfile (this is how mine is set up too). It seems like env files are still picked up by a build at run time.

**Comment by EnablingLogic** [comment]
I'm also having this issue. `.env` and `.env.production` are copied to `.next/standalone`, but `.env.local` is missing.

**Comment by mikgross** [comment]
Hey all, same for .env.test

**Comment by tolgaand** [comment]
same issue, any updates?

**Comment by dogukanakkaya** [comment]
I ran into the same issue with my GitHub Actions CI. I've created the env file in the CI itself:
```
 - name: "Create frontend env file"
    run: |
          touch apps/web/.env
          echo NEXT_PUBLIC_ENV1=${{ secrets.NEXT_PUBLIC_ENV1 }} >> apps/web/.env
          echo NEXT_PUBLIC_ENV2=${{ secrets.NEXT_PUBLIC_ENV2 }} >> apps/web/.env
```

**Comment by Dacaramo** [comment]
This is still an issue in 2024. If you do not precede all of the variables names with NEXT_PUBLIC the variables will not be available on the server. This leaves the SSR feature of storing secrets on the Next.js server completely unusable. This must be fixed as soon as possible.

**It is important to notice that this is an issue not only for environment variables defined in the .env file, but also for environment variables that are passed to next build command.**

Doing:

```ENV
MY_VARIABLE=my-variable-value next build
```

Does not work neither, using `process.env.MY_VARIABLE` server side will always give undefined.

**Comment by kuhlaid** [comment]
I'm having a problem simply getting my app to read updated environment variables. I have no idea where NextJs is caching the environment. I'm using `yarn dev` to test locally, and NextJs says it is reading my `.env.development` environment file when the server starts, but my `process.env` values do not change. I deleted the .next directory from the app., I moved any other .env files out of the root directory. I logged out of my computer and logged back in, and of course the correct environment values began appearing in my app. When in doubt restart your computer. :P

**Comment by dfilipov-glitnor** [comment]
We hit this issue when building our containers in Github Actions.
The `.env` and `.env.production` files are imported but not the `.env.local` 
We're currently copying that file manually in the container as a workaround, but this is creating unneeded technical debt.

**Comment by alivarastepour** [comment]
My `.env.production` file is not being loaded in the production build neither. I had to add my variables to `.env` to get it working.

**Comment by duannx** [comment]
Nextjs only copies `.env` to the `standalone` folder so we need to manually copy other environment files (`.env.local`, `.env.production`... ) to the `standalone` folder as well to make them work together.

**Comment by ViniciusCestarii** [comment]
> Nextjs only copies .env to the standalone folder so we need to manually copy other environment files (.env.local, .env.production... ) to the standalone folder as well to make them work together.

This is so confusing and should at least be mentioned on the docs.

Environment variables should be loaded at runtime rather than being bundled into the build or included in the standalone folder.

**Comment by aldinokemal** [comment]
@ViniciusCestarii Yeah, that confused me too. In Remix, we don't need the environment variables during the build process. I migrated from Remix to Next, and it's been a bit perplexing.

**Comment by esmalleydev** [comment]
bump... found thing wondering why .env.local wasnt working >.>

**Comment by duannx** [comment]
> This is so confusing and should at least be mentioned on the docs.
> 
> Environment variables should be loaded at runtime rather than being bundled into the build or included in the standalone folder.

There are two types of environment variables in Next.js.  

The first type is runtime environment variables, which are only available in Node.js. These are loaded at runtime.  

The second type is `NEXT_PUBLIC_*` variables, [which are inlined into the JavaScript bundle](https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#bundling-environment-variables-for-the-browser) at build time and delivered to the client. This is why some variables can't be changed after building for production.  

We usually deploy the `standalone` folder to our server only, so Next.js has to copy `.env.*` files to the standalone output. It does copy `.env` and `.env.production`, but not `.env.local` (I was wrong about `.env.production` in my previous comment—updated it). This makes sense, as `.env` and `.env.production` are meant for production use. However, this behavior should be clearly documented in the official documentation.

**Comment by masto** [comment]
This is massively confusing, and I have gone around in circles for the past 3 hours because all of the _documentation_, and many comments on forums, github issues, stackoverflow, etc., say that _on the server_, env variables are dynamic. That is **not** matching my experience today, as a new Next.js developer on 15.2.0 with the app router, a standalone build, and attempting to use the current, up-to-date versions and ways of doing everything.

As best as I have been able to tell, @Dacaramo 's comment from August 2024 https://github.com/vercel/next.js/issues/46296#issuecomment-2313365313 is the only one anywhere that actually mentions this doesn't work, and it seems to still be the only accurate description. All my experimentation seems to indicate that the environment is being baked in during the standalone build, and cannot be influenced afterward through .env.anything files, process env variables, or otherwise.

This leaves me wondering exactly how one is meant to build a standalone Docker image that needs things like API keys, database endpoints, or various other server-side configuration provided by the user, in the way that is typically done through the environment in Docker deployments.

Maybe I am totally confused and missing the obvious. I am certainly missing sleep trying to puzzle this out. I think what I am inferring is that few people actually deploy Next.js apps this way, so the problem goes mostly unnoticed and unfixed?

**Comment by hood** [comment]
There is also an issue where NEXT_PUBLIC_ env variables are not getting inlined at all for client components in standalone deployments, making them totally impossible to retrieve. https://github.com/vercel/next.js/issues/80194

**Comment by karthicedric7-cloud** [comment]
This is a classic env drift issue variables present locally but missing in the built output. I built a CLI that catches this before you deploy: "npx safelaunch scan" it checks for missing vars, .env.local sync issues, and standalone build misconfigs. Might help anyone hitting this.

**Comment by marcoshernanz** [comment]
Thanks for the report. This is expected behavior for standalone output. Next.js copies `.env` and `.env.production`, but intentionally excludes `.env.*.local` files. Pass server-only values to the container at runtime with `--env` or `--env-file`, or explicitly copy the local env file next to `server.js`. `NEXT_PUBLIC_*` values are inlined at build time. This is the same behavior covered in #59130, so I’m closing this issue.
